### PR TITLE
feat: parse arithmetic operations

### PR DESCRIPTION
**Pull request by Marbjo07**
Improved tokenizer, and AST builder
Parses:
- Addition, Subtraction, Division, Multiplication and Exponentiation

All current tests are passing, indicating that basic linear functions are possible input.

### CODE REVIEW

**Comment by Marbjo07** [comment]
There should be created tests for functions including division and exponentiation. These should also pass before merging.

**Comment by Marbjo07** [comment]
I think that we shouldn't push the content of testdata. This data is generated using `go generate ./...`, it should be simple to add to the build process. However I think it's easier done after merging with main.

**Comment by Marbjo07** [comment]
> There should be created tests for functions including division and exponentiation. These should also pass before merging.

Test cases are made and are passing

**Comment by Marbjo07** [comment]
There are currently a lot of repetition. Think its better done in another pr

**Comment by Yuqzii** [comment]
Why is the testdata json files? Is it not simpler and more readable to just have it in the _test file?

File: internal/guessTheFunction/expr.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+func (Add) isExpr() {}
+
+func (Subtract) isExpr() {}
+
+func (Multiply) isExpr() {}
+
+func (Divide) isExpr() {}
+
+func (Power) isExpr() {}
```
Comment (Yuqzii) [inline internal/guessTheFunction/expr.go:83]: What is isExpr() used for?

File: internal/guessTheFunction/ast.go [hunk 2]
```diff
@@ -1,61 +1,154 @@
...
-}
-
-func (m Multiply) Eval(x float64) float64 {
-	return m.Left.Eval(x) * m.Right.Eval(x)
-}
-
-func (d Divide) Eval(x float64) float64 {
-	return d.Left.Eval(x) / d.Right.Eval(x)
-}
+func BuildAST(tokens []Token) (Expr, error) {
```
Comment (Yuqzii) [inline internal/guessTheFunction/ast.go:145]: Could this not be private? I'm assuming it is never needed outside the guessTheFunction package.

File: internal/guessTheFunction/expr.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+type Divide struct {
+	Left  Expr `json:"left"`
+	Right Expr `json:"right"`
+}
+
+type Power struct {
+	Base     Expr `json:"base"`
+	Exponent Expr `json:"exponent"`
+}
```
Comment (Yuqzii) [inline internal/guessTheFunction/expr.go:41]: Does all this also need to be public?

**Review by Yuqzii — COMMENTED** [review]
Generally just make things that does not need to be public private.

File: internal/guessTheFunction/guessTheFunction.go [hunk 4]
```diff
@@ -1,12 +1,52 @@
...
 	"log"
 
 	"github.com/bwmarrin/discordgo"
 )
 
+// Defined here so that it can be accessed outside the module
+type TestCase struct {
+	Input    string `json:"input"`
+	Expected Expr   `json:"expected"`
+}
```
[the remark this answers is not in the document]
Comment (Marbjo07) [reply internal/guessTheFunction/guessTheFunction.go:15]: It was previously defined in guessTheFunction_test.go, where it was not accessible. Yea I meant package and it's used in the test generation tool.

**Comment by Marbjo07** [comment]
To explain my intention of creating .json files for the test data. 
- First, I think it would be easier to debug, test and validate, if we generate random test data in the future.
- We needed to have some functionality for saving and loading functions. This can be used to, for example, prevent a server crash from losing data about a running game/round.  

This is just the solution I think makes the most sense. What do you (@Yuqzii) think?

**Comment by Yuqzii** [comment]
The json loading seems pretty useful, but I don't think randomly generated test data is a good idea. It makes it difficult to guarantee that edge cases are checked, and makes testing inconsistent and more difficult to debug if the tests are not always the same.
Loading from json to prevent data loss is smart, and we should totally keep it for that, but I think more manual test data is better. And with manual test data it might as well be inside the _test file.

**Comment by Marbjo07** [comment]
I agree, maybe random test data isn't optimal. We could use some template system? 
For example if we define a function `[Number]x^2 + [Number]x + [Number]` [Number] could then be replaced with edgecases of numbers, for example large negatives, fractions, decimals, and other expressions that doesn't include `x`
I think with this approach that we still can cover all edgecases.

**Comment by Yuqzii** [comment]
Yes, but I still think it will just take longer to make a system that creates these automatically, and they should just be generated once anyways (so that debugging is easier). AI can be pretty decent for generating test data as well, and is probably easier than creating a program to generate them.

File: internal/guessTheFunction/guessTheFunction.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Yuqzii) [reply internal/guessTheFunction/guessTheFunction.go]: Weird, should be accessible inside the entire package as long as it is defined in one file in that package. Doesn't really matter though.

[the remark this answers is not in the document]
Comment (Marbjo07) [reply internal/guessTheFunction/guessTheFunction.go]: It's an easy fix, but according to the docs. `... give that file a name ending in "_test.go". The file will be excluded from regular package builds ...` https://pkg.go.dev/testing

File: internal/guessTheFunction/expr.go (hunk 1 above)
Comment (Marbjo07) [reply internal/guessTheFunction/expr.go]: `isExpr()` is just a marker to make sure only the types we want can be used as expr. It ensures only the types defined can be used with the expr interface. But when I think of it, Eval should satisfy this.

**Comment by Marbjo07** [comment]
Ok, it makes sense to just write the test cases. I think my proposed changes should be better. 
All the testcases are defined in `_testcases`. But im unsure if keeping `expr_test.go` makes any sense. It could just be moved into `guessTheFunction_test.go`

**Comment by Marbjo07** [comment]
I think we are ready to merge.

File: internal/guessTheFunction/expr.go (hunk 1 above)
Comment (Yuqzii) [reply internal/guessTheFunction/expr.go]: Ok, so just to make it clear that they implement the interface?

**Review by Yuqzii — APPROVED** [review]

**Comment by Yuqzii** [comment]
Delete the branch?
